### PR TITLE
--terragrunt-modules-that-include filter for run-all

### PR DESCRIPTION
**Pull request by yorinasub17**
## Motivation

With the introduction of multiple include blocks, it is becoming increasingly common to setup terragrunt projects that have shared configurations for specific components. However, this introduces a problem for infrastructure CI/CD pipelines, where you want to only run against the modules that changed.

Consider our example in [terragrunt-infrastructure-live-example](https://github.com/gruntwork-io/terragrunt-infrastructure-live-example). The common configurations related to `mysql` are extracted in the file `_envcommon/mysql.hcl`. If we were bumping the module version for the `qa` environment by editing the source reference in [non-prod/us-east-1/qa/mysql/terragrunt.hcl](https://github.com/gruntwork-io/terragrunt-infrastructure-live-example/blob/master/non-prod/us-east-1/qa/mysql/terragrunt.hcl#L10), the pipeline is straightforward - we just run `terragrunt plan` and `terragrunt apply` in the folder of the file that changed.

However, what should we do if the common configuration gets updated? E.g., bumping the instance type: https://github.com/gruntwork-io/terragrunt-infrastructure-live-example/blob/master/_envcommon/mysql.hcl#L40

It isn't immediately obvious what the desired behavior is in this case. For example, for something like changing the instance type of the database, the desired behavior is probably to run `terragrunt plan` and `terragrunt apply` progressively across the environments. That is, start with `qa`, then `stage`, and then `prod`. On the other hand, for a change like updating the environment variables for an application, you can get away with running `terragrunt plan` and `terragrunt apply` in parallel across all environments.

This PR seeks to provide a basic building block that can be used for designing a pipeline for both use cases.

## Solution

In this PR, we introduce a flag for `run-all` (`--terragrunt-modules-that-include`) which will filter out all modules in a stack that does not include at least one of the files listed in the arg.

With such a filter, you can implement the two flows:

For the promotion flow, you can have your pipeline run each environment in succession:
```
terragrunt run-all plan --terragrunt-modules-that-include ../_envcommon/mysql.hcl --terragrunt-working-dir non-prod/qa
# if that succeeds
terragrunt run-all plan --terragrunt-modules-that-include ../_envcommon/mysql.hcl --terragrunt-working-dir non-prod/stage
# ... and so on ...
```

For the parallel flow, you can call `run-all` from the root once:
```
terragrunt run-all plan --terragrunt-modules-that-include _envcommon/mysql.hcl
```

Refer to the docs update in this PR for more details on the use case and how this feature can help implement the pipeline.

### CODE REVIEW

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md [hunk 1]
```diff
@@ -409,3 +411,34 @@ terraform {
...
+were updated, you can run `terragrunt plan` or `terragrunt apply` by passing in the updated file with
+`--terragrunt-config`.
+
+However, if you use `include` or `read_terragrunt_config`, then a single file change may need to be reflected on
+multiple files that were not touched at all in the commit. In our previous example, when a configuration is updated in
+the `_env/app.hcl` file, we need to apply the change to all the modules that `include` that common environment
+configuration.
+
+Terragrunt currently does not have any features for supporting this use case when `read_terragrunt_config` is
+used. However, for `include` blocks, you can use the
```
Comment (brikis98) [inline docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:432]: Should we have something for `read_terragrunt_config`? Should `--terragrunt-modules-that-include` be something like `--terragrunt-modules-that-reference` instead, and support all possible ways to include or reference some other file?

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md [hunk 2]
```diff
@@ -409,3 +411,34 @@ terraform {
...
+[--terragrunt-modules-that-include]({{site.baseurl}}/docs/reference/cli-options/#terragrunt-modules-that-include) CLI
+option for the `run-all` command.
+
+In the previous example, your CI/CD pipeline can run `terragrunt run-all plan --terragrunt-modules-that-include
+_env/app.hcl`. This will:
+
+- Recursively find all Terragrunt modules in the current directory tree.
+- Filter out any modules that don't include `_env/app.hcl` so that they won't be touched.
+- Run `plan` on any modules remaining (which will be the set of modules in the current tree that include
```
Comment (brikis98) [inline docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:441]: How does `apply` work in this case? In particular:

1. If multiple modules in a single env include something that was modified: e.g., you have some `_env/db.hcl` and you use that 5 times in the `dev` environment to deploy 5 DBs. Will / should our CI / CD pipeline try to `apply` them all in parallel? I'm guessing the answer is yes.
2. If multiple modules across different envs include this thing: e.g., the `_env/db.hcl` is used in `dev`, `stage`, and `prod`. Will / should our CI / CD pipeline try to `apply` them all in parallel? I'm guessing the answer here should be no. But then, is our pipeline able to run `plan` in one env, prompt for approval, if approved, run `apply`, then automatically repeat the process with each subsequent env?

File: docs/_docs/04_reference/cli-options.md [hunk 3]
```diff
@@ -735,6 +736,8 @@ When passed it, sets logging level for terragrunt. All supported levels are:
 
 **CLI Arg**: `--terragrunt-check`<br/>
 **Environment Variable**: `TERRAGRUNT_CHECK` (set to `true`)
+**Commands**:
+- [hclfmt](#hclfmt)
```
Comment (brikis98) [inline docs/_docs/04_reference/cli-options.md:740]: Is this related to this PR?

File: docs/_docs/04_reference/cli-options.md [hunk 4]
```diff
@@ -762,5 +767,65 @@ block by specifying `<BLOCK>.<ATTR>`, where `<BLOCK>` is the block name: e.g., `
...
+- [plan-all (DEPRECATED: use run-all)](#plan-all-deprecated-use-run-all)
+- [apply-all (DEPRECATED: use run-all)](#apply-all-deprecated-use-run-all)
+- [output-all (DEPRECATED: use run-all)](#output-all-deprecated-use-run-all)
+- [destroy-all (DEPRECATED: use run-all)](#destroy-all-deprecated-use-run-all)
+- [validate-all (DEPRECATED: use run-all)](#validate-all-deprecated-use-run-all)
+
+When passed in, `run-all` will only run the command against Terragrunt modules that include the specified file.
+
+This applies to the set of modules that are identified based on all the existing criteria for deciding which modules to
+include. For example, consider the following folder structure:
```
Comment (brikis98) [inline docs/_docs/04_reference/cli-options.md:791]: NIT: You show the file structure here, but not the contents of the files, so it's not obvious what uses `include` on what if you're not already familiar with Terragrunt.

**Review by brikis98 — DISMISSED** [review]
Overall, code changes LGTM! Just had some questions / NITs that can be answered separately from this PR.

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 1 above)
Comment (yorinasub17) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:432]: Ideally yes, but it's non-trivial to implement this using the tools that the `hcl` library gives us, since function calls can't be parsed out. That is, there is no way to say "give me all the calls to `read_terragrunt_config`," like there is a way to say "give me all the `include` blocks."

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 2 above)
Comment (yorinasub17) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:441]: I mentioned this in the PR description, but realized I didn't do it for the docs so expanded on it: 9c4dc14b3fae49d361e230796c3890420e353dec

Note that I think this is still flawed, in that the commit is a one way street. That is, if you abort after deploying to `stage`, you are now in an awkward state where you have to roll this change out when you want to touch `prod` since the updates are already merged for `qa` and `stage`.

I think the real solution here is going to be figuring out how to implement remote `include`, although that has its own can of worms as we discussed.

File: docs/_docs/04_reference/cli-options.md (hunk 3 above)
Comment (yorinasub17) [reply docs/_docs/04_reference/cli-options.md:740]: Yes, and it comes from the format change in https://github.com/gruntwork-io/terragrunt/pull/1893/files/fe655623b8730eeb428e6588fe1f2a48b557767e#diff-0ff83c0cbafb0d058b828ba1777066af4bc6eee4f726cecc0e0351b8f15ee65eR776

File: docs/_docs/04_reference/cli-options.md (hunk 4 above)
Comment (yorinasub17) [reply docs/_docs/04_reference/cli-options.md:791]: Makes sense! Added in example: 8eb213b4567d56ed6f2fdd8dfb807414a280a2ae

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 1 above)
Comment (brikis98) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:432]: I'm going to toss out an idea that may be dumb, but just in case: since we implement the `read_terragrunt_config` helper, could it maintain an internal cache (map) of (a) which files were read in and (b) which files were doing the reading? If so, could that cache be used to do the `--terragrunt-modules-that-include` style filtering?

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 2 above)
Comment (brikis98) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:441]: Ah, yea, you're right. This would require a rollback if the deploy failed part way through. I think the best we can do for now is document the need for rollback, both in the Terragrunt docs, and our own pipeline docs. In the future, as you said, remote `include` is prob the proper solution.

**Review by brikis98 — DISMISSED** [review]
Two comments left, neither blocking, so giving it a "ship it" if you want to merge/release this and come back to the other items later.

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 1 above)
Comment (yorinasub17) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:432]: I think that would work, but I am a bit hesitant with that approach primarily because it is essentially relying on a global variable that gets updated with functions that have side effects. It feels... messy...

I filed https://github.com/gruntwork-io/terragrunt/issues/1910 so we don't lose track of it.

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 2 above)
Comment (yorinasub17) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:441]: Good point about documenting. What do you think about adcafb8f3b8656794301dec00c90355c94f3bdf6?

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 1 above)
Comment (brikis98) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:432]: Roger, thx!

File: docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md (hunk 2 above)
Comment (brikis98) [reply docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md:441]: LGTM

**Review by brikis98 — APPROVED** [review]
LGTM

**Comment by yorinasub17** [comment]
Thanks for review and working through this! I'm going to merge this in now.
